docs/moderate.md

```diff
@@ -0,0 +1,193 @@
+```
+
+Creating the trigger function that changes the status of the comment based on the result of the openai API.
+
+```sql
+CREATE OR REPLACE FUNCTION moderate_comment() RETURNS TRIGGER AS $$
+declare
+    out jsonb;
+BEGIN
+  select openai_moderate(
+    'text-moderation-stable',
+    NEW.body
+  )->'results'->0 into out;
+  NEW.status = get_moderation_status(out);
+
+  RETURN NEW;
+END;
+$$ LANGUAGE plpgsql;
+```
+
+Creating the trigger:
+
+```sql
+CREATE TRIGGER moderate_comment_trigger
+BEFORE INSERT OR UPDATE ON comments
+  FOR EACH ROW EXECUTE FUNCTION moderate_comment();
+```
+
+Testing the trigger:
+
+```sql
+insert into comments (body) values
+  ('I love the new product'),
+  ('He is an asshole'),
+  ('I want to kill them all');
+```
+
+Checking the results:
+
+```sql
+table comments;
+id |          body           |         created_at         |   status
+----+-------------------------+----------------------------+------------
+  1 | I love the new product  | 2024-06-07 19:07:10.884519 | approved
+  2 | He is an asshole        | 2024-06-07 19:07:10.884519 | harassment
+  3 | I want to kill them all | 2024-06-07 19:07:10.884519 | violence
+```
+
+## Via background action
+
+Background options will not be blocking your transactions, so it's a better option for
+moderating comments in a production environment and for a large number of comments.
+
+Check out [background_actions.md](background_actions.md) for more information on
+how to setup background actions to use open ai keys properly.
+
+You can get the full example in the [bg_worker_moderate.sql](examples/bg_worker_moderate.sql).
+
+For the background action, instead of the trigger, we will create a procedure
+that will moderate the comments:
+
+```sql
+CREATE OR REPLACE FUNCTION get_moderation_status(body TEXT, api_key TEXT)
+RETURNS TEXT AS $$
+DECLARE
+    result JSONB;
+    category TEXT;
+BEGIN
+    -- Call OpenAI moderation endpoint
+    select openai_moderate( 'text-moderation-stable', body, api_key)->'results'->0 into result;
+
+    -- Check if any category is flagged
+    IF result->>'flagged' = 'true' THEN
+        FOR category IN SELECT jsonb_object_keys(result->'categories') LOOP
+            IF (result->'categories'->>category)::BOOLEAN THEN
+                RETURN category;
+            END IF;
+        END LOOP;
+    END IF;
+
+    RETURN 'approved';
+END;
+$$ LANGUAGE plpgsql;
+```
+
+We'll also create a procedure that will moderate the comments:
+
+```sql
+CREATE OR REPLACE PROCEDURE check_new_comments_to_moderate(job_id int, config jsonb) LANGUAGE PLPGSQL AS
+$$
+declare
+  comment record;
+  api_key text;
+BEGIN
+  RAISE NOTICE 'Executing action % with config %', job_id, config;
+  -- iterate over comments and moderate them
+  api_key := config->>'api_key';
+  for comment in select * from comments where status = 'pending' loop
+    update comments set status = get_moderation_status(comment.body, api_key)
+    where id = comment.id;
+  end loop;
```

Comment:
same comment about the "select for update skip locked" technique

Reply:
was updating based on the learnings! very good!